### PR TITLE
check escalation for executor subject

### LINKED ISSUE #153: check the executor subject permission for action apply
this is a sub-task of https://github.com/open-cluster-management-io/work/issues/124 , we need to check the executor subject permission for action apply

proposal link: https://github.com/open-cluster-management-io/enhancements/tree/main/enhancements/sig-architecture/34-work-executor-group

### PR DESCRIPTION
**Pull request by zhujian7**
Signed-off-by: zhujian <jiazhu@redhat.com>
To fix https://github.com/open-cluster-management-io/work/issues/153#issuecomment-1248855226

### CODE REVIEW

File: test/integration/executor_test.go [hunk 1]
```diff
@@ -209,4 +209,160 @@ var _ = ginkgo.Describe("ManifestWork Executor Subject", func() {
...
+			roleName := "role1"
+			_, err = spokeKubeClient.RbacV1().ClusterRoles().Create(
+				context.TODO(), &rbacv1.ClusterRole{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: roleName,
+					},
+					Rules: []rbacv1.PolicyRule{
+						{
+							// with "escalate" verb
+							// Verbs: []string{"create", "update", "patch", "get", "list", "delete", "escalate"},
```
Comment (zhujian7) [inline test/integration/executor_test.go:313]: it is weird that `Verbs: []string{"create", "update", "patch", "get", "list", "delete", "escalate"},` does not work, but `Verbs: []string{"*"},` does

**Comment by zhujian7** [comment]
cc @qiujian16 @elgnay

File: pkg/spoke/controllers/manifestcontroller/manifestwork_controller.go [hunk 2]
```diff
@@ -84,7 +86,7 @@ func NewManifestWorkController(
 		hubHash:                   hubHash,
 		restMapper:                restMapper,
 		appliers:                  apply.NewAppliers(spokeDynamicClient, spokeKubeClient, spokeAPIExtensionClient),
-		validator:                 auth.NewExecutorValidator(spokeKubeClient),
+		validator:                 auth.NewExecutorValidator(spokeKubeConfig, spokeKubeClient),
```
Comment (qiujian16) [inline pkg/spoke/controllers/manifestcontroller/manifestwork_controller.go]: I think you may want to init this outside of the controller, you can reuse it when delete later.

File: pkg/spoke/auth/auth.go [hunk 3]
```diff
@@ -97,15 +117,61 @@ func (v *sarValidator) Validate(ctx context.Context, executor *workapiv1.Manifes
...
+	dynamicClient, err := v.newImpersonateClientFunc(v.config, username(sa.Namespace, sa.Name))
+	if err != nil {
+		return err
+	}
+
+	_, err = dynamicClient.Resource(gvr).Namespace(namespace).Create(ctx, obj, metav1.CreateOptions{
+		DryRun: []string{"All"},
+	})
+	if apierrors.IsAlreadyExists(err) {
+		return nil
```
Comment (elgnay) [inline pkg/spoke/auth/auth.go]: If the k8s resource exists and its spec is different from the required one, we may need to update the k8s resource and see if the user has the permission or not. What do you think?

Comment (zhujian7) [reply pkg/spoke/auth/auth.go]: I checked even if the k8s resource exists, the API server will return the forbidden error when receiving an escalated request.
And I added 2 integration tests to cover these.

**Pull request merged**